core/shared/src/main/scala/org/typelevel/log4cats/PagingSelfAwareStructuredLogger.scala

```diff
@@ -0,0 +1,223 @@
+
+  private class PagingSelfAwareStructuredLogger[F[_]: Monad](pageSizeK: Int, maxPageNeeded: Int)(
+      sl: SelfAwareStructuredLogger[F]
+  ) extends SelfAwareStructuredLogger[F] {
+    assert(pageSizeK > 0)
+    assert(maxPageNeeded > 0)
+
+    private val pageIndices = (1 to maxPageNeeded).toList
+    private val logSplitIdN = "log_split_id"
+    private val pageSize = pageSizeK * 1024
+
+    private def pagedLogging(
+        loggingOp: (=> String) => F[Unit],
+        logSplitId: String,
+        msg: => String
+    ): F[Unit] = {
+      val numOfPagesRaw = (msg.length - 1) / pageSize + 1
+      val numOfPages = Math.min(numOfPagesRaw, maxPageNeeded)
+      if (numOfPages <= 1)
+        loggingOp(msg)
+      else {
+        val logSplitIdPart1 = logSplitId.split('-').head
+        val pageHeaderTail = s"$numOfPages $logSplitIdPart1"
+        val pageFooterTail = s"$numOfPages $logSplitIdN=$logSplitId page_size=$pageSizeK Kb"
+        pageIndices
+          .take(numOfPages)
+          .traverse_ { pi =>
+            val beginIndex = (pi - 1) * pageSize
+            val pageContent = msg.slice(beginIndex, beginIndex + pageSize)
+
+            loggingOp(show"""Page $pi/$pageHeaderTail
+                            |
+                            |$pageContent
+                            |
+                            |Page $pi/$pageFooterTail""".stripMargin)
+          }
+      }
+    }
+
+    private def addCtx(
+        msg: => String,
+        ctx: Map[String, String]
+    ): (String, Map[String, String]) = {
+      val logSplitId = UUID.randomUUID().show
+      (
+        logSplitId,
+        ctx
+          .updated(logSplitIdN, logSplitId)
+          .updated("page_size", s"${pageSizeK.show} Kb")
+          .updated("log_size", s"${msg.length.show} Byte")
+      )
+    }
+
+    @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
+    private def doLogging(
+        loggingLevelChk: => F[Boolean],
+        logOpWithCtx: Map[String, String] => (=> String) => F[Unit],
+        msg: => String,
+        ctx: Map[String, String] = Map()
+    ): F[Unit] = {
+      loggingLevelChk.ifM(
+        {
+          val (logSplitId, newCtx) = addCtx(msg, ctx)
+          pagedLogging(logOpWithCtx(newCtx), logSplitId, msg)
+        },
+        Applicative[F].unit
+      )
+    }
+
+    @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
+    private def doLoggingThrowable(
+        loggingLevelChk: => F[Boolean],
+        logOpWithCtx: Map[String, String] => (=> String) => F[Unit],
+        t: Throwable,
+        msg: => String,
+        ctx: Map[String, String] = Map()
+    ): F[Unit] = {
+      loggingLevelChk.ifM(
+        doLogging(loggingLevelChk, logOpWithCtx, s"$msg\n${getStackTrace(t)}", ctx),
+        Applicative[F].unit
+      )
+    }
+
+    def getStackTrace(t: Throwable): String = {
+      val bos = new ByteArrayOutputStream()
+      val ps = new PrintStream(bos)
+      t.printStackTrace(ps)
+      bos.toString
```

Comment:
This depends on the system's default character set.  You can circumvent all those nasty character encoding issues by using `Writer`s instead of `Stream`s.  Here's the commons-lang implementation in Java.  It's _almost_ the same as yours.

```java
   public static String getStackTrace(final Throwable throwable) {
        final StringWriter sw = new StringWriter();
        final PrintWriter pw = new PrintWriter(sw, true);
        throwable.printStackTrace(pw);
        return sw.getBuffer().toString();
    }
```

Reply:
Took it, thanks.